### ISSUE TITLE
Functional test screen_basic_spec failing

### ISSUE THREAD

**Issue by Numkil** [neovim/neovim#2130]
Operating System:  Fedora 21

This  test fails 1  time out of every 4 runs or so.

![screenshot from 2015-03-11 08 12 34](https://cloud.githubusercontent.com/assets/5604686/6591891/8a4802ca-c7c6-11e4-8430-59907af07c0a.png)
![screenshot from 2015-03-11 08 11 56](https://cloud.githubusercontent.com/assets/5604686/6591890/8a45fc46-c7c6-11e4-8ec9-7fb2ff9b1df0.png)

If I can help in any other way, please do not hesitate to ask.

**Comment by Pyrohh** [comment]
I've experience this off and on, but I've no idea how to reproduce it. It definitely happened before #2029 was merged.

**Comment by Numkil** [comment]
I ran the test 200 times as an experiment and captured the amount of times it failed. 
11 out of 200 tests failed, each time the same reason as the one in the picture above.

**Comment by Pyrohh** [comment]
Did you just run them via a for loop in your shell, and append the exit codes to a log file? I can't imagine this would be too hard to bisect, unless you're already doing that.

**Comment by doppioandante** [comment]
I wanted to look at this bug, but I can't even reproduce it :| 
Does it depend from the used shell?

**Comment by Pyrohh** [comment]
I've noticed this on mksh, if that matters.

**Comment by Numkil** [comment]
Yes I just used a for loop in my shell. I'm using bash, with terminator as the emulator.

**Comment by doppioandante** [comment]
Unfortunately I can't reproduce it, no matter what.
I found out that the terminal emulator doesn't make any difference, because the gui being tested is specifically written for testing il lua.
I tried to follow the resizing codepath, and the only point where I think something could go wrong is: https://github.com/neovim/neovim/blob/master/src/nvim/screen.c#L7161
Another possibility is lua or the testing infrastructure iself, but I'd rule those out.
If you can still reproduce the bug, I'll put here a patch to perform some logging, if you don't mind running it. Maybe that can shed some light.

**Comment by Numkil** [comment]
Sure if you share the patch, I will run it for you. Just tell me what info is important for you.

**Comment by Numkil** [comment]
I'm not able to reproduce the failure anymore. This is the current commit I'm at. 
![screenshot from 2015-03-20 08 14 36](https://cloud.githubusercontent.com/assets/5604686/6747400/2b61d14a-ced9-11e4-960d-768a6becb245.png)

I suspect the error may have been caused by something from my system itself? I run fedora so updates are frequent.

**Comment by justinmk** [comment]
ok, we will leave it at that unless it returns. Thanks

**Comment by bfredl** [comment]
I'm still seeing this occassionally. `while (make test) do; done` typically fails after 4-5 iterations. I will look more into debugging this later.

**Comment by bfredl** [comment]
Hmm, It seems that `notification_cb` might be called 1-2 times more after `stop()` is called ( in `Screen:wait`). So the correct output is seen, and stop is called, but `notification_cb` is run once more overwriting `err` with something which might be a failure...

**Comment by bfredl** [comment]
Seems to fix it, but dunno if it is the correct way to fix it:

    diff --git a/test/functional/ui/screen.lua b/test/functional/ui/screen.lua
    index f79d634..0f1352f 100644
    --- a/test/functional/ui/screen.lua
    +++ b/test/functional/ui/screen.lua
    @@ -216,19 +216,22 @@ function Screen:expect(expected, attr_ids, attr_ignore)
     end
     
     function Screen:wait(check, timeout)
    -  local err, checked = false
    +  local err, checked, stopped = false
       local function notification_cb(method, args)
         assert(method == 'redraw')
         self:_redraw(args)
         err = check()
         checked = true
         if not err then
    +      stopped = true
           stop()
         end
         return true
       end
       run(nil, notification_cb, nil, timeout or default_screen_timeout)
    -  if not checked then
    +  if stopped then
    +    return
    +  elseif not checked then
         err = check()
       end
       if err then
    -- 
    2.3.0

ping @tarruda Should `stop()` be indeterministic in this way?

**Comment by bfredl** [comment]
did a full [capture of all redraw events](https://gist.github.com/bfredl/3f98375656b26f204d71). Please look near the end; it is clear it "bounces" between the correct output, and the incorrect `resize^` output. I'll push PR to fix the robustness of the screen test runner, but it seems this could be a glitch of the resize redrawing in `nvim` as well.

**Comment by tarruda** [comment]
> did a full capture of all redraw events. Please look near the end; it is clear it "bounces" between the correct output, and the incorrect resize^ output. I'll push PR to fix the robustness of the screen test runner, but it seems this could be a glitch of the resize redrawing in nvim as well.

This bouncing probably happens because we need a "semaphore"(call to `screen:except` or `wait`) to wait for a certain state before proceeding with the test input. It happened to @justinmk in #1802([ref](https://github.com/neovim/neovim/pull/1802#issuecomment-74940079)) but I remember having to fix this in other tests.

Here's a simple(but real) example the illustrates the problem:

```lua
    it('`yes` and is interrupted with CTRL-C', function()
      feed(':call system("yes")<cr>')
      feed('<c-c>')
      screen:expect([[
        ^                                                     |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        Type  :quit<Enter>  to exit Vim                      |
      ]])
    end)
```

The above test executes the "yes" command and immediately sends a ctrl+c command to interrupt it. Depending on which state the editor is when these commands arrive, the ctrl+c may cause the `system` command to not even be executed(Ctrl+c will kill all unprocessed data in the input buffer). Here's the correct test:

```lua
    it('`yes` and is interrupted with CTRL-C', function()
      feed(':call system("yes")<cr>')
      screen:expect([[
                                                             |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        :call system("yes")                                  |
      ]])
      feed('<c-c>')
      screen:expect([[
        ^                                                     |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        Type  :quit<Enter>  to exit Vim                      |
      ]])
    end)
```

This version fixes the problem by waiting for the command to be executed before sending the ctrl+c. Here's a small variant that can cause the "bouncing" problem:

```lua
    it('`yes` and is interrupted with CTRL-C', function()
      feed('<c-c>')
      feed(':call system("yes")<cr>')
      feed('<c-c>')
      screen:expect([[
        ^                                                     |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        ~                                                    |
        Type  :quit<Enter>  to exit Vim                      |
      ]])
    end)
```

This one first sends a ctrl+c, resulting in the "Type :quit..." message being displayed and then executes a command which will remove the message. Since the final expectation is the "Type :quit" message, its possible that the `screen:expect` call will fail after receiving the expected screen state(because there are pending changes). Again, the problem is fixed by waiting for the correct state after each action.

**Comment by bfredl** [comment]
@tarruda if I understand you correctly there should be an `screen:expect` before the `try_resize(1,1`)` as well? But I still don't see what is causing the intermediately correct state _before_ the resize call...

**Comment by tarruda** [comment]
> @tarruda if I understand you correctly there should be an screen:expect before the try_resize(1,1)` as well? But I still don't see what is causing the intermediately correct state before the resize call...

I will look into it

**Comment by bfredl** [comment]
but what do you think of #2221? I think it is still correct, as we are anyway not trying to detect the "final" state, just that some redrawing state shows the expected output.

**Comment by tarruda** [comment]
I cant reproduce this locally, but maybe this patch will help:

```lua
diff --git a/test/functional/ui/screen_basic_spec.lua b/test/functional/ui/screen_basic_spec.lua
index 034e9a0..306bec6 100644
--- a/test/functional/ui/screen_basic_spec.lua
+++ b/test/functional/ui/screen_basic_spec.lua
@@ -453,7 +453,7 @@ describe('Screen', function()
       feed('iresize')
     end)

-    it('rebuilds the whole screen', function()
+    local function rebuilds_whole_screen()
       screen:expect([[
         resize^                   |
         ~                        |
@@ -461,9 +461,12 @@ describe('Screen', function()
         ~                        |
         -- INSERT --             |
       ]])
-    end)
+    end
+
+    it('rebuilds the whole screen', rebuilds_whole_screen)

     it('has minimum width/height values', function()
+      rebuilds_whole_screen()
       screen:try_resize(1, 1)
       screen:expect([[
         -- INS^ERT --|

```

**Comment by tarruda** [comment]
This is simpler and may also fix the race conditition:

```lua
diff --git a/test/functional/ui/screen_basic_spec.lua b/test/functional/ui/screen_basic_spec.lua
index 034e9a0..6ee8e95 100644
--- a/test/functional/ui/screen_basic_spec.lua
+++ b/test/functional/ui/screen_basic_spec.lua
@@ -1,7 +1,7 @@
 local helpers = require('test.functional.helpers')
 local Screen = require('test.functional.ui.screen')
 local clear, feed, execute = helpers.clear, helpers.feed, helpers.execute
-local insert = helpers.insert
+local insert, wait = helpers.insert, helpers.wait

 describe('Screen', function()
   local screen
@@ -451,6 +451,7 @@ describe('Screen', function()
     before_each(function()
       screen:try_resize(25, 5)
       feed('iresize')
+      wait()
     end)

     it('rebuilds the whole screen', function()

```

**Comment by bfredl** [comment]
yes, the `wait()` removes the indetermininsm :+1:

**Comment by tarruda** [comment]
The thing is that `feed('iresize')` will return as soon as the data is in the input buffer, but it may be unprocessed when the `try_resize` call is made, creating the indeterminism.

**Comment by bfredl** [comment]
That seems to be it, using `redraw_debug()` from #2221 it looks like `wait()` eleminates this indetermininsm and the resulting "bounce".

**Comment by tarruda** [comment]
> but what do you think of #2221? I think it is still correct, as we are anyway not trying to detect the "final" state, just that some redrawing state shows the expected output.

Even so, I think we should leave the possibility of bouncing. Think of it as a "trap" to detect wrong test assumptions(in this case the assumption was that try_resize waits for the previous input). Though it may be better to detect the bouncing and print a message that suggests the inclusion of expect/wait after input in the middle of the test.

**Comment by bfredl** [comment]
>  Though it may be better to detect the bouncing and print a message that suggests the inclusion of expect/wait after input in the middle of the test.

:+1: I will try change #2221 to do this instead.

**Comment by bfredl** [comment]
Should be fixed by #2221, so closing.
